src/coreclr/jit/async.cpp

```diff
@@ -836,330 +1196,387 @@ void Async2Transformation::Transform(
+//   object.
+//
+// Parameters:
+//   resumeByteArrLclNum - Local that has the continuation object's data array
+//   liveLocals          - Information about each live local.
+//   resumeBB            - Basic block to append IR to
+//
+void Async2Transformation::RestoreFromDataOnResumption(unsigned                             resumeByteArrLclNum,
+                                                       const jitstd::vector<LiveLocalInfo>& liveLocals,
+                                                       BasicBlock*                          resumeBB)
+{
+    // Copy data
+    for (const LiveLocalInfo& inf : liveLocals)
     {
-        resumeObjectArrLclNum = GetGCDataArrayVar();
+        if (inf.DataSize <= 0)
+        {
+            continue;
+        }
 
-        newContinuation       = m_comp->gtNewLclvNode(m_comp->lvaAsyncContinuationArg, TYP_REF);
-        unsigned gcDataOffset = m_comp->info.compCompHnd->getFieldOffset(m_async2Info.continuationGCDataFldHnd);
-        GenTree* gcDataInd    = LoadFromOffset(newContinuation, gcDataOffset, TYP_REF);
-        GenTree* storeAllocedObjectArr = m_comp->gtNewStoreLclVarNode(resumeObjectArrLclNum, gcDataInd);
-        LIR::AsRange(resumeBB).InsertAtEnd(LIR::SeqTree(m_comp, storeAllocedObjectArr));
+        LclVarDsc* dsc = m_comp->lvaGetDesc(inf.LclNum);
+
+        GenTree* byteArr = m_comp->gtNewLclvNode(resumeByteArrLclNum, TYP_REF);
+        unsigned offset  = OFFSETOF__CORINFO_Array__data + inf.DataOffset;
+        GenTree* cns     = m_comp->gtNewIconNode((ssize_t)offset, TYP_I_IMPL);
+        GenTree* addr    = m_comp->gtNewOperNode(GT_ADD, TYP_BYREF, byteArr, cns);
 
-        // Copy GC pointers
-        for (LiveLocalInfo& inf : m_liveLocals)
+        GenTree* value;
+        if ((dsc->TypeGet() == TYP_STRUCT) || dsc->IsImplicitByRef())
         {
-            if (inf.GCDataCount <= 0)
-            {
-                continue;
-            }
+            value = m_comp->gtNewBlkIndir(dsc->GetLayout(), addr, GTF_IND_NONFAULTING);
+        }
+        else
+        {
+            value = m_comp->gtNewIndir(dsc->TypeGet(), addr, GTF_IND_NONFAULTING);
+        }
 
-            LclVarDsc* dsc = m_comp->lvaGetDesc(inf.LclNum);
-            if (dsc->TypeGet() == TYP_REF)
-            {
-                GenTree* objectArr = m_comp->gtNewLclvNode(resumeObjectArrLclNum, TYP_REF);
-                unsigned offset    = OFFSETOF__CORINFO_Array__data + (inf.GCDataIndex * TARGET_POINTER_SIZE);
-                GenTree* value     = LoadFromOffset(objectArr, offset, TYP_REF);
-                GenTree* store     = m_comp->gtNewStoreLclVarNode(inf.LclNum, value);
+        GenTree* store;
+        if (dsc->IsImplicitByRef())
+        {
+            GenTree* baseAddr = m_comp->gtNewLclvNode(inf.LclNum, dsc->TypeGet());
+            // TODO-CQ: Incoming data has no non-null GC refs, so this does not need write barriers.
+            // Backend does not handle GTF_IND_TGT_NOT_HEAP for STORE_BLK.
```

Comment:
Besides, this is actually `NOT_HEAP`, right?

Reply:
Yes, let me update this. It's a bit subtle because the local can be an implicit byref parameter that we actually type as `TYP_BYREF` today, even though they are always stack-only (this issue is https://github.com/dotnet/runtime/issues/82166).